### DISCUSSION TITLE
Conditional WHERE clause do not compile

### DISCUSSION THREAD

**Discussion by artkoenig** [simolus3/drift#1543] [Q&A]
Hi, I get a compiler error with the following query:

```
  Stream<int> streamNewTransactionsCountForAccount(int? accountId) {
    final count = coalesce<int>([transactions. isNew.count(), const Constant(0)]);
    final query = selectOnly(transactions)
      ..addColumns([count, transactions.isNew, transactions.accountId])
      ..where(transactions.isNew.equals(true) & (accountId != null ? transactions.accountId.equals(accountId) : const Constant(true)));
    return query.map((row) => row.read(count)).watchSingle();
  }

```
The error for the second condition in the WHERE clause is: `The argument type 'FunctionParameter' can't be assigned to the parameter type 'Expression<bool?>'.`

but this is ok:

```
  Stream<int> streamNewTransactionsCountForAccount(int? accountId) {
    final count = coalesce<int>([transactions. isNew.count(), const Constant(0)]);
    final query = selectOnly(transactions)
      ..addColumns([count, transactions.isNew, transactions.accountId])
      ..where(transactions.isNew.equals(true) & (accountId != null ? transactions.accountId.isIn([accountId]) : const Constant(true)));
    return query.map((row) => row.read(count)).watchSingle();
  }
```
do I miss something?

**Comment by simolus3** [comment]
One thing that stands out is that `isIn` returns an `Expression<bool?>` whereas `equals` returns an `Expression<bool>`. The `&` operator is defined for `Expression<bool>` which _shouldn't_ be a problem since `Expression<bool>` is a subtype of `Expression<bool?>`. However, it looks like type inference doesn't like that at all.

Another example of this:

```dart
Expression<bool> x = accountId != null ? transactions.accountId.equals(accountId) : const Constant(true); // fine
Expression<bool?> y = x; // fine too
Expression<bool?> z = accountId != null ? transactions.accountId.equals(accountId) : const Constant(true); // compile-time error
```

I'm actually a bit surprised about Dart behaving this way, but that's the way it is. You can fix this by making the type parameter explicit: `const Constant<bool>(true)`.
